contracts/modules/staking/StakingPool.sol

```diff
@@ -580,81 +580,75 @@ contract StakingPool is IStakingPool, ERC721 {
-    // trancheCount = maxTrancheId - firstTrancheIdToUse + 1
+    uint firstTrancheIdToUse = (block.timestamp + request.period + request.gracePeriod) / TRANCHE_DURATION;
     uint trancheCount = block.timestamp / TRANCHE_DURATION + MAX_ACTIVE_TRANCHES - firstTrancheIdToUse;
+    uint remainingAmount = Math.divCeil(request.amount, NXM_PER_ALLOCATION_UNIT);
 
     (
-      uint[] memory trancheAllocatedCapacities,
-      uint initialCapacityUsed
-    ) = getAllocatedCapacities(
-      request.productId,
-      firstTrancheIdToUse,
-      trancheCount
-    );
-
-    (uint[] memory totalCapacities, uint totalCapacity) = getTotalCapacitiesForTranches(
+      uint[] memory trancheAllocations,
+      uint[] memory trancheCapacities,
+      uint requestedTranchesCapacityUsed,
+      uint requestedTranchesCapacity
+    ) = getTrancheAllocationsAndCapacities(
       request.productId,
       firstTrancheIdToUse,
       trancheCount,
       request.globalCapacityRatio,
-      request.capacityReductionRatio
-    );
-
-    uint remainingAmount = Math.divCeil(request.amount, NXM_PER_ALLOCATION_UNIT);
-    // total capacity can get below the used capacity as a result of burns
-    require(
-      totalCapacity > initialCapacityUsed && totalCapacity - initialCapacityUsed >= remainingAmount,
-      "StakingPool: Insufficient capacity"
+      request.capacityReductionRatio,
+      remainingAmount
     );
 
     {
       uint[] memory coverTrancheAllocation = new uint[](trancheCount);
 
       for (uint i = 0; i < trancheCount; i++) {
 
-        if (trancheAllocatedCapacities[i] >= totalCapacities[i]) {
+        if (trancheAllocations[i] >= trancheCapacities[i]) {
           continue;
         }
 
-        uint availableTrancheCapacity = totalCapacities[i] - trancheAllocatedCapacities[i];
-        uint allocate = Math.min(availableTrancheCapacity, remainingAmount);
+        uint allocate = Math.min(trancheCapacities[i] - trancheAllocations[i], remainingAmount);
 
         remainingAmount -= allocate;
         allocatedCoverAmount += allocate;
-        trancheAllocatedCapacities[i] += allocate;
+        trancheAllocations[i] += allocate;
         coverTrancheAllocation[i] = allocate;
 
         if (remainingAmount == 0) {
           break;
         }
-
-        // technically should never happen because of the total-used capacity check
-        require(remainingAmount == 0, "StakingPool: Insufficient capacity");
       }
 
-      updateAllocatedCapacities(
+      // technically should never happen because of the initial capacity check
+      require(remainingAmount == 0, "StakingPool: Insufficient capacity");
+
+      updateAllocations(
         request.productId,
         firstTrancheIdToUse,
         trancheCount,
-        trancheAllocatedCapacities
+        trancheAllocations
       );
 
+      uint targetBucketId =
+        (block.timestamp + request.period + request.gracePeriod)
+        / BUCKET_DURATION
+        + 1;
```

Comment:
why is this + 1 ?

Is it because it's expiring the bucket after the bucker in which gracePeriod ends?